File: src/baseliner/checks/registry.py
from __future__ import annotations

from baseliner.checks.base import Check


class CheckRegistry:
    def __init__(self) -> None:
        self._registry: dict[str, Check] = {}

    def register(self, check: Check) -> None:
        self._registry[check.check_id] = check

    def get(self, check_id: str) -> Check:
        if check_id not in self._registry:
            raise KeyError(f"Unknown check id: '{check_id}'")
        return self._registry[check_id]

    def all(self) -> dict[str, Check]:
        return dict(self._registry)


def build_default_registry() -> CheckRegistry:
    from baseliner.checks.git import DefaultBranchIsMain, StaleRepo
    from baseliner.checks.hygiene import (
        CiPresent,
        CodeownersExists,
        DependencyUpdateConfig,
        GitignoreExists,
        LicenseExists,
        ReadmeExists,
        ReadmeHasHeading,
        ReadmeNonEmpty,
    )

    registry = CheckRegistry()
    for check in [
        ReadmeExists(),
        ReadmeNonEmpty(),
        ReadmeHasHeading(),
        LicenseExists(),
        GitignoreExists(),
        CiPresent(),
        CodeownersExists(),
        DependencyUpdateConfig(),
        DefaultBranchIsMain(),
        StaleRepo(),
    ]:
        registry.register(check)
    return registry

File: internal/checks/registry.go
package checks

// Registry maps check IDs to Check implementations.
type Registry struct {
	m map[string]Check
}

// NewRegistry returns an empty registry.
func NewRegistry() *Registry {
	return &Registry{m: make(map[string]Check)}
}

// Register adds a check, keyed by its ID.
func (r *Registry) Register(c Check) {
	r.m[c.ID()] = c
}

// Get returns the check for an ID, and whether it was found.
func (r *Registry) Get(id string) (Check, bool) {
	c, ok := r.m[id]
	return c, ok
}

// BuildDefault returns a registry with all 10 built-in checks registered.
func BuildDefault() *Registry {
	r := NewRegistry()
	for _, c := range []Check{
		readmeExists{base{"readme_exists", LayerFS}},
		readmeNonEmpty{base{"readme_nonempty", LayerFS}},
		readmeHasHeading{base{"readme_has_heading", LayerFS}},
		licenseExists{base{"license_exists", LayerFS}},
		gitignoreExists{base{"gitignore_exists", LayerFS}},
		ciPresent{base{"ci_present", LayerFS}},
		codeownersExists{base{"codeowners_exists", LayerFS}},
		dependencyUpdateConfig{base{"dependency_update_config", LayerFS}},
		defaultBranchIsMain{base{"default_branch_is_main", LayerGit}},
		staleRepo{base{"stale_repo", LayerGit}},
	} {
		r.Register(c)
	}
	return r
}
